src/buffer/bytes.rs

```diff
@@ -1,163 +1,118 @@
-                write!(f, "Deallocation::Foreign {{ capacity: unknown }}")
-            }
-        }
-    }
-}
-
-/// A continuous, fixed-size, immutable memory region that knows how to de-allocate itself.
+/// A continuous memory region that may be allocated externally.
 ///
-/// In the most common case, this buffer is allocated using [`allocate_aligned`](alloc::allocate_aligned)
-/// and deallocated accordingly [`free_aligned`](alloc::free_aligned).
-/// When the region is allocated by a foreign allocator, [Deallocation::Foreign], this calls the
-/// foreign deallocator to deallocate the region when it is no longer needed.
-pub struct Bytes<T: NativeType> {
-    /// inner data
-    data: MaybeForeign<T>,
-    /// how to deallocate this region
-    deallocation: Deallocation,
+/// In the most common case, this is created from [`Vec`].
+/// However, this region can also be allocated by a foreign allocator.
+pub struct Bytes<T> {
+    /// An implementation using an `enum` of a `Vec` or a foreign pointer is not used
+    /// because `deref` is at least 50% more expensive than the deref of a `Vec`.
+    data: ManuallyDrop<Vec<T>>,
+    /// the region was allocated
+    allocation: Allocation,
 }
 
-impl<T: NativeType> Bytes<T> {
-    /// Takes ownership of an allocated memory region,
-    ///
-    /// # Arguments
-    ///
-    /// * `ptr` - Pointer to raw parts
-    /// * `len` - Length of raw parts in **bytes**
-    /// * `capacity` - Total allocated memory for the pointer `ptr`, in **bytes**
-    ///
+impl<T> Bytes<T> {
+    /// Takes ownership of an allocated memory region `[ptr, ptr+len[`,
     /// # Safety
-    ///
-    /// This function is unsafe as there is no guarantee that the given pointer is valid for `len`
-    /// bytes. If the `ptr` and `capacity` come from a `Buffer`, then this is guaranteed.
-    ///
-    /// # Panics
-    ///
-    /// This function panics if the give deallocation not is `Deallocation::Foreign`
+    /// This function is safe iff:
+    /// * the region is properly allocated in that a slice can be safely built from it.
+    /// * the region is immutable.
+    /// # Implementation
+    /// This function leaks iff `owner` does not deallocate the region when dropped.
     #[inline]
-    pub unsafe fn from_ffi(
+    pub unsafe fn from_owned(
         ptr: std::ptr::NonNull<T>,
         len: usize,
-        deallocation: Deallocation,
+        owner: Arc<dyn RefUnwindSafe + Send + Sync>,
     ) -> Self {
-        assert!(matches!(deallocation, Deallocation::Foreign(_)));
         // This line is technically outside the assumptions of `Vec::from_raw_parts`, since
         // `ptr` was not allocated by `Vec`. However, one of the invariants of this struct
         // is that we do not expose this region as a `Vec`; we only use `Vec` on it to provide
         // immutable access to the region (via `Vec::deref` to `&[T]`).
+        // MIRI does not complain, which seems to agree with the line of thought.
```

Comment:
Maybe we could also use `Box<[T]>` and call `Vec::from_raw_parts` only on `get_vec` on native allocated data.

Reply:
I tried this but in `get_vec` we would be creating a temporary Vec and returning `&mut Vec`, which does not compile :(

Reply:
Oh yeah.. :(